### PR TITLE
options=values for GET /v2/types operation

### PR DESCRIPTION
**Pull request by fgalan**
Implements #1170 

@crbrox 
@kzangeli

### CODE REVIEW

File: src/lib/mongoBackend/mongoQueryTypes.cpp [hunk 1]
```diff
@@ -149,6 +149,126 @@ static long long countEntities(const std::string& tenant, const std::vector<std:
...
+(
+  EntityTypeVectorResponse*            responseP,
+  const std::string&                   tenant,
+  const std::vector<std::string>&      servicePathV,
+  std::map<std::string, std::string>&  uriParams,
+  unsigned int*                        totalTypesP
+)
+{
+  unsigned int   offset         = atoi(uriParams[URI_PARAM_PAGINATION_OFFSET].c_str());
+  unsigned int   limit          = atoi(uriParams[URI_PARAM_PAGINATION_LIMIT].c_str());
```
Comment (kzangeli) [inline src/lib/mongoBackend/mongoQueryTypes.cpp]: Do we have any check anywhere to verify that offset and limit are actually integers?

File: src/lib/orionTypes/EntityTypeVectorResponse.cpp [hunk 2]
```diff
@@ -129,16 +129,24 @@ void EntityTypeVectorResponse::release(void)
 */
 std::string EntityTypeVectorResponse::toJson(ConnectionInfo* ciP)
 {
-  std::string  out = "[";
+  std::string  out = "[";
```
Comment (kzangeli) [inline src/lib/orionTypes/EntityTypeVectorResponse.cpp]: space at EOL

File: src/lib/serviceRoutinesV2/getEntityAllTypes.cpp [hunk 3]
```diff
@@ -64,7 +64,14 @@ std::string getEntityAllTypes
     totalTypesP = &totalTypes;
   }
 
-  TIMED_MONGO(mongoEntityTypes(&response, ciP->tenant, ciP->servicePathV, ciP->uriParam, ciP->apiVersion, totalTypesP));
+  if (ciP->uriParamOptions["values"])
+  {
+    TIMED_MONGO(mongoEntityTypesValues(&response, ciP->tenant, ciP->servicePathV, ciP->uriParam, totalTypesP));
+  }
+  else  // defaul
```
Comment (kzangeli) [inline src/lib/serviceRoutinesV2/getEntityAllTypes.cpp]: default

File: test/functionalTest/cases/1027_get_all_entity_types/get_all_entity_types.test [hunk 4]
```diff
@@ -38,6 +38,7 @@ brokerStart CB
 # 06. GET /v2/types and get a list of three types, each with one entity
 # 07. Create E4/T3/A4+A3
 # 08. GET /v2/types and get a list of three types, T3 now has two entities
+# 09. GET /v2/types wiht limit, getting 2 types but Fiware-Total-Count 3
```
Comment (kzangeli) [inline test/functionalTest/cases/1027_get_all_entity_types/get_all_entity_types.test]: typo: wiht

I imagine this test is implemented already but was not in the 'TOC' ?

File: test/functionalTest/cases/1170_get_types_values_option/get_types_values_option.test [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+--SHELL--
+
+#
+# 01. GET /v2/types?options=values and get an empty list
+# 02. Create E1/T1/A1+A2
+# 03. GET /v2/types?options=values and get a list with one element
+# 04. Create E2/T2/A1+A3
+# 05. Create E3/T3/A2+A4
+# 06. GET /v2/types?options=values and get a list of three types
+# 07. GET /v2/types?options=values wiht limit, getting 2 types but Fiware-Total-Count 3
```
Comment (kzangeli) [inline test/functionalTest/cases/1170_get_types_values_option/get_types_values_option.test]: copy 'n paste: wiht

File: test/functionalTest/cases/1170_get_types_values_option/get_types_values_option_with_null_type_entities.test [hunk 6]
```diff
@@ -0,0 +1,258 @@
...
+
+#
+# 01. GET /v2/types?options=values and get an empty list
+# 02. Create E1/T1/A1+A2
+# 03. GET /v2/types?options=values and get a list with one element
+# 04. Create E2/<null>/A1+A3 (using NGSIv1)
+# 05. Create E3/T3/A2+A4
+# 06. GET /v2/types?options=values and get a list of three types
+# 07. GET /v2/types?options=values wiht limit, getting 2 types but Fiware-Total-Count 3
+# 08. GET /v2/types?options=values wiht offset=3, thus out of limits
```
Comment (kzangeli) [inline test/functionalTest/cases/1170_get_types_values_option/get_types_values_option_with_null_type_entities.test]: Plenty of wiht in the functests ...

File: src/lib/mongoBackend/mongoQueryTypes.cpp (hunk 1 above)
Comment (fgalan) [inline src/lib/mongoBackend/mongoQueryTypes.cpp]: I haven't paid too much attention to that piece of code in this PR. Note that it is the same that used for mongoEntityTypes() so I understand that if it was ok that time, then it is ok this time.

File: test/functionalTest/cases/1027_get_all_entity_types/get_all_entity_types.test (hunk 4 above)
Comment (fgalan) [inline test/functionalTest/cases/1027_get_all_entity_types/get_all_entity_types.test]: Yes, the fix was align TOC with the already implemented step.

File: test/functionalTest/cases/1170_get_types_values_option/get_types_values_option_with_null_type_entities.test (hunk 6 above)
Comment (fgalan) [inline test/functionalTest/cases/1170_get_types_values_option/get_types_values_option_with_null_type_entities.test]: Overall replacement done in c117f01ef3e6d64fc68acdbd5220823f16f3965a

File: test/functionalTest/cases/1170_get_types_values_option/get_types_values_option.test (hunk 5 above)
Comment (fgalan) [inline test/functionalTest/cases/1170_get_types_values_option/get_types_values_option.test]: Fixed in c117f01ef3e6d64fc68acdbd5220823f16f3965a

File: test/functionalTest/cases/1027_get_all_entity_types/get_all_entity_types.test (hunk 4 above)
Comment (fgalan) [inline test/functionalTest/cases/1027_get_all_entity_types/get_all_entity_types.test]: wiht fixed in c117f01ef3e6d64fc68acdbd5220823f16f3965a

File: src/lib/serviceRoutinesV2/getEntityAllTypes.cpp (hunk 3 above)
Comment (fgalan) [inline src/lib/serviceRoutinesV2/getEntityAllTypes.cpp]: Fixed in c117f01ef3e6d64fc68acdbd5220823f16f3965a

File: src/lib/orionTypes/EntityTypeVectorResponse.cpp (hunk 2 above)
Comment (fgalan) [inline src/lib/orionTypes/EntityTypeVectorResponse.cpp]: c117f01ef3e6d64fc68acdbd5220823f16f3965a

File: src/lib/mongoBackend/mongoQueryTypes.cpp (hunk 1 above)
Comment (kzangeli) [inline src/lib/mongoBackend/mongoQueryTypes.cpp]: It's not very import and and it has not much to do really with this PR.
Some day we should look into it though.
Also, why have the `atoi(uriParams[URI_PARAM_PAGINATION_OFFSET].c_str()` in more than one place ... (should be str2double or something like that).
Creating an issue.

NTC

**Comment by kzangeli** [comment]
LGTM

**Comment by crbrox** [comment]
LGTM

**Pull request merged**